src/Microsoft.OData.Client/ALinq/GroupByAnalyzer.cs

```diff
@@ -0,0 +1,529 @@
+            if (keySelector.Body.NodeType != ExpressionType.MemberInit && keySelector.Body.NodeType != ExpressionType.New)
+            {
+                return false;
+            }
+
+            // Scenario 3: GroupBy(d1 => new { d1.Property1, ..., d1.PropertyN })
+            // Scenario 4: GroupBy(d1 => new Cls { ClsProperty1 = d1.Property1, ..., ClsPropertyN = d1.PropertyN }) - not common but possible
+            GroupByKeySelectorAnalyzer.Analyze(input, keySelector);
+
+            return true;
+        }
+
+        /// <summary>
+        /// Ensure apply query option for the resource set is initialized
+        /// </summary>
+        /// <param name="input">The resource expression</param>
+        private static void EnsureApplyInitialized(QueryableResourceExpression input)
+        {
+            Debug.Assert(input != null, $"{nameof(input)} != null");
+
+            if (input.Apply == null)
+            {
+                ResourceBinder.AddSequenceQueryOption(input, new ApplyQueryOptionExpression(input.Type));
+            }
+        }
+
+        private static bool TryBindToInput(ResourceExpression input, LambdaExpression le, out Expression bound)
+        {
+            List<ResourceExpression> referencedInputs = new List<ResourceExpression>();
+            bound = InputBinder.Bind(le.Body, input, le.Parameters[0], referencedInputs);
+
+            if (referencedInputs.Count > 1 || (referencedInputs.Count == 1 && referencedInputs[0] != input))
+            {
+                bound = null;
+            }
+
+            return bound != null;
+        }
+
+        /// <summary>
+        /// Analyzes a GroupBy key selector for property or properties that the input sequence is grouped by.
+        /// </summary>
+        private sealed class GroupByKeySelectorAnalyzer : DataServiceALinqExpressionVisitor
+        {
+            /// <summary>The input resource, as a queryable resource.</summary>
+            private readonly QueryableResourceExpression input;
+
+            /// <summary>The key selector lambda parameter.</summary>
+            private readonly ParameterExpression lambdaParameter;
+
+            /// <summary>
+            /// Creates an <see cref="GroupByKeySelectorAnalyzer"/> expression.
+            /// </summary>
+            /// <param name="source">The source expression.</param>
+            /// <param name="paramExpr">The parameter of the key selector expression, e.g., d1 => new { d1.ProductId, d1.CustomerId }.</param>
+            private GroupByKeySelectorAnalyzer(QueryableResourceExpression source, ParameterExpression paramExpr)
+            {
+                this.input = source;
+                this.lambdaParameter = paramExpr;
+            }
+
+            /// <summary>
+            /// Analyzes a GroupBy key selector for property or properties that the input sequence is grouped by. 
+            /// </summary>
+            /// <param name="input">The input resource expression.</param>

```

Comment:
I think having examples of these expressions in these doc strings will make it very easy for someone to understand what is going on.

Reply:
@ElizabethOkerio Given an expression like:
```csharp
dataServiceContext.Sales.GroupBy(
    d1 => d1.ProductId,
    (d2, d3) => new { ProductId = d2, AverageAmount = d3.Average(d4 => d4.Amount) })
```
the input/source expression is the expression representing the `Sales` resource set.
About the usage of the word `source` or `input` it's based on the `GroupBy` method definition:
```csharp
GroupBy<TSource, TKey, TResult>(
    IQueryable<TSource> source,
    Expression<Func<TSource, TKey>> keySelector,
    Expression<Func<TKey, IEnumerable<TSource>, TResult>> resultSelector)
```
Providing this expression as an example whenever I refer to `QueryableResourceExpression` would be a tough call so what I have gone ahead and done is improve the doc and make it more consistent:
```csharp
/// <param name="input">The resource set expression (source) for the GroupBy.</param>
```